CI/CD

Sidduppal opened the pull request:
:mechanic: Unit tests for `prodigal.py` and `samtools.py`
Both are in preliminary stage. Any critique and comments on how to improve are welcome.

Sidduppal committed:
Resolves #43
code currently in the if __name__ == '__main__': block moved to main
Output filehandle writes lines directly as they are being read from output file
Replaced manual creatio of temporary files with tempfile module

Sidduppal committed:
Resolves Issue #55
- Removed dispatcher dictionary by using globals()
- Changed the location of checking TypeError in get_versions() function

Sidduppal committed:
Revert "Resolves #43"
Done so that future commits are not added on the same PR relating to hmmer.py

This reverts commit e2c8bc1cae94e97582d1ed35a812db0562a5ab6c.

Sidduppal committed:
Revert "Resolves Issue #55"
Done to make sure that future commits are not added to the PR relating to hmmer.py

This reverts commit fc582de91db9a6e76e5a38291f858e0c4a7a3daf.

Sidduppal committed:
Resolves #43
code currently in the if __name__ == '__main__': block moved to main
    Output filehandle writes lines directly as they are being read from output file
    Replaced manual creatio of temporary files with tempfile module

Sidduppal committed:
Resolves #43

Sidduppal committed:
Resolves #43
Added the funtionality to filter the results in case the user already has the hmmscan table

Sidduppal committed:
changed st.size to path.getsize

Co-authored-by: Evan Rees <25933122+WiscEvan@users.noreply.github.com>

Sidduppal committed:
Resolves #43
changed st.size to path.getsize

Sidduppal committed:
Merge branch 'dev' of https://github.com/KwanLab/Autometa into dev

Sidduppal committed:
Merge branch 'dev' of https://github.com/KwanLab/Autometa into dev

Sidduppal committed:
Merge branch 'dev' of https://github.com/KwanLab/Autometa into dev

Sidduppal committed:
Merge branch 'CI/CD' of https://github.com/WiscEvan/Autometa into CI_CD

Sidduppal committed:
unit test for prodigal

Sidduppal committed:
unit test for samtools.py

Sidduppal committed:
Unit test for prodigal.py

Sidduppal committed:
Unit test for coverage.py

Sidduppal committed:
Unit test for diamond.py

Sidduppal committed:
Merge branch 'dev' of https://github.com/KwanLab/Autometa into CI_CD

Sidduppal committed:
mocking prodigal run

Sidduppal commented:
One thing I wanted to ask you was that to test functions `contigs_from_headers` and `orf_records_from_contigs` I think we'll need to upload an amino acid output file generated from prodigal run. This file is used as an argument for these 2 functions. I can l mock the files but then we would not able to able to test any of the output generated by the functions.

Sidduppal committed:
Added fully functional unit tests (with mocking) for coverage.py

Sidduppal committed:
Miscellaneous additions
:art: Working tests for prodigal.py
:art: Improvements in exceptions.py and prodigal.py

Sidduppal committed:
Unit test for kmers.py
:art: Minor edits in kmers.py
:art: Additions in test_coverage.py

Sidduppal committed:
pulling dev

Sidduppal committed:
merging from dev

Sidduppal committed:
Unit tests for kmers.py
:white_check_mark: unit test for kmers.count(), kmers.normalize() and kmers.embed()
:art: Minor changes in kemrs.py

Sidduppal committed:
Merge branch 'dev' of https://github.com/WiscEvan/Autometa into CI_CD

Sidduppal committed:
Merge branch 'CI/CD' of https://github.com/WiscEvan/Autometa into CI_CD

Sidduppal committed:
Updated more unit tests
:white_check_mark: Units tests for metagenome class, mg.length_filter(), mg.call_orfs()
:art: Now uses tmpdir_factory
:art: minor changes in metagenome.py

evanroyrees commented on autometa/common/coverage.py:
```diff
@@ -212,11 +213,11 @@ def align_reads(
                 break
 
-        if (not fwd_reads or not rev_reads) and step == "full":
+        if ((not fwd_reads or not rev_reads) and not se_reads) and step == "full":
```
Single end reads are not necessarily required.

Sidduppal replied:
I added this because there could be a case when the user just provided the `se_reads` and not the `fwd_reads` and `rev_reads`. If I removed, ` and not se_reads` then it will throw an error even if the user-provided `se_reads`

evanroyrees replied:
I have not tried this with only single-end reads. Do you know if the calculation sequence will work with only single-end reads?

evanroyrees commented on autometa/common/coverage.py:
```diff
@@ -170,7 +171,7 @@ def get(
         return from_spades_names(records=[rec for rec in SeqIO.parse(fasta, "fasta")])
 
-    try:
+    with tempfile.TemporaryDirectory() as tempdir:
```
Do we want to remove files if the run fails? How will the user be able to troubleshoot the error?
```suggestion
    try:
```

Sidduppal replied:
But isn't putting `finally` at the end of the `try and except` block doing the same thing? I noticed that the `finally` statememt is executed even if an error is raised, so isn't `finally` also removing the `tempdir` even if the tool fails?

evanroyrees commented on autometa/common/coverage.py:
```diff
@@ -226,8 +227,6 @@ def align_reads(
             if calculation.__name__ == "parse_bed":
                 return calculation()
             calculation()
-    finally:
-        shutil.rmtree(tempdir, ignore_errors=True)
```
See comment above

evanroyrees commented on autometa/common/exceptions.py:
```diff
@@ -102,6 +102,7 @@ def __str__(self):
         """
         return f"{self.value}\n{message}"
 
+
```
Not sure why this is here

evanroyrees commented on autometa/common/external/prodigal.py:
```diff
@@ -185,7 +185,7 @@ def run(assembly, nucls_out, prots_out, force=False, cpus=0, parallel=True):
             with open(fp) as fh:
                 for _ in SimpleFastaParser(fh):
                     pass
-        except (IOError, ValueError):
+        except (OSError, ValueError):
```
Does OSError result from `SimpleFastaParser`?

Sidduppal replied:
I was unable to replicate the error. But I must have seen it being raised instead of `IOError` and that's why I would have put it there. Moreover, from [this](https://stackoverflow.com/a/29347946/12671809) discussion it seems we should be using `OSError` instead of `IOError` as `IOError` is now just an alias for `OSError`

Sidduppal replied:
Unfortunately BioPython has not specified the errors its packages are raising in its [documentation](https://biopython.org/docs/1.75/api/Bio.SeqIO.FastaIO.html#Bio.SeqIO.FastaIO.SimpleFastaParser).
If you want we can catch both `IOError1 and `OSError`

evanroyrees replied:
We should probably raise `OSError` rather than `IOError` as well.

evanroyrees commented on autometa/common/kmers.py:
```diff
@@ -82,11 +82,13 @@ def init_kmers(kmer_size=5):
 
     """
-    kmers = {}
+    if not float(kmer_size).is_integer():
+        raise TypeError(f"kmer_size must be an int! Given: {type(kmer_size)}")
+    kmer_size = int(float(kmer_size))
```
Let's simplify this and not try any conversion. Seems like a fair enough ask from the user.

Also, if we wanted, we could start to use [type hints](https://docs.python.org/3/library/typing.html "type hints python page") as well.

```suggestion
    if not isinstance(kmer_size, int):
        raise TypeError(f"kmer_size must be an int! Given: {type(kmer_size)}")
```

evanroyrees commented on autometa/common/kmers.py:
```diff
@@ -127,7 +129,7 @@ def load(kmers_fpath):
     try:
         df = pd.read_csv(kmers_fpath, sep="\t", index_col="contig")
-    except ValueError:
+    except (ValueError, TypeError):
```
Is this another error that occurs from `pd.read_csv(...)`?

Sidduppal replied:
Yeah, I encountered this when I was passing in a Mock or MagicMock object to it. It seems it is raised when the input file is not a correct object

evanroyrees replied:
It sounds like this should be handled separately, since the error is a result from a different input. I.e. this error is not a result of the `contig` column being absent.

evanroyrees commented on autometa/common/kmers.py:
```diff
@@ -225,7 +227,7 @@ def seq_counter(assembly, ref_kmers, verbose=True):
     -------
     dict
-        {contig:[count,count,...]} count index is respective to ref_kmers.keys()
+        {contig:[count,count,...]} count index is respective to ref_kmers keys()
```
This was showing the `dict.keys()` method, `ref_kmers.keys()`

```suggestion
        {contig:[count,count,...]} count index is respective to ref_kmers.keys()
```

evanroyrees commented on autometa/common/kmers.py:
```diff
@@ -315,7 +317,8 @@ def count(
         logger.warning(f"counts already exist: {out} force to overwrite. [retrieving]")
         df = pd.read_csv(out, sep="\t", index_col="contig")
     else:
-        if not isinstance(size, int):
+        # checks if it is something like 12.0 vs. 12.9. Also check is an int
+        if not float(size).is_integer():
```
We only care that size is an integer. We do not want to perform any conversion.

`if not isinstance(size, int)` seems to cover all of our bases here.

evanroyrees commented on autometa/common/kmers.py:
```diff
@@ -495,23 +498,25 @@ def embed(
         try:
             df = pd.read_csv(kmers, sep="\t", index_col="contig")
-        except ValueError:
+        except (ValueError, TypeError):
```
Does `TypeError` get raised from `pd.read_csv(...)`?

Sidduppal replied:
Yeah, I encountered this when I was passing in a Mock or MagicMock object to it. It seems it is raised when the input file is not a correct object

evanroyrees replied:
See comment above

evanroyrees commented on autometa/common/kmers.py:
```diff
@@ -495,23 +498,25 @@ def embed(
         try:
             return pd.read_csv(out, sep="\t", index_col="contig")
-        except ValueError:
+        except (ValueError, TypeError):
```
Does `TypeError` get raised from `pd.read_csv(...)`?

evanroyrees replied:
See comment above

evanroyrees commented on autometa/common/kmers.py:
```diff
@@ -495,23 +498,25 @@ def embed(
         embed_desc = f"out:{out} type:{type(out)}"
-        requirements = f"kmers type must be a pd.DataFrame or filepath."
+        requirements = f"Given pd.DataFrame is empty!"
+        # requirements = f"kmers type must be a pd.DataFrame or filepath."
```
```suggestion

```

evanroyrees commented on autometa/common/metagenome.py:
```diff
@@ -357,11 +357,11 @@ def length_filter(self, out, cutoff=3000, force=False):
-            # https://stackoverflow.com/a/4187220/13118765
-            raise TypeError(f"cutoff: {cutoff} must be a float or int")
+        try:
+            cutoff = float(cutoff)
+        except ValueError:
+            raise ValueError(f"Cutoff should be flaot of int, specified {cutoff}")
```
This looks like we are replacing a more robust type check with a brute force conversion. I think we should leave the former

evanroyrees reviewed (changes requested):
Just comments on the non-test files for now

evanroyrees commented on tests/unit_tests/test_metagenome.py:
```diff
@@ -1,89 +1,92 @@
-    }
-    # taxa_df = test_metagenome.assign_taxonomy(force=False, kwargs=params)
-    # assert taxa_df.shape == (3587, )
+    nucls_fp, prots_fp = init_Metagenome.call_orfs()
```
`self.call_orfs()` has three parameters that should be tested here.

### 😍 🤖 💚  Loving the use of patching prodigal 👍 

Would we be able to do introspection here to ensure the `prodigal.run(...)` command looks as what we would expect when we call `self.call_orfs(...)`?

#### Args
* force
* cpus
* parallel

i.e.
```python
with pytest.raises(TypeError):
    nucls_fp, prots_fp = init_Metagenome.call_orfs(force="false")

with pytest.raises(TypeError):
    nucls_fp, prots_fp = init_Metagenome.call_orfs(parallel="F")

with pytest.raises(TypeError):
    nucls_fp, prots_fp = init_Metagenome.call_orfs(cpus=4.0)
```

### What about when the ORFs files already exist?
We also need to ensure the `self.prot_orfs_fpath` and `self.nucl_orfs_fpath` are returned when the ORFs files already exist.

evanroyrees commented on tests/unit_tests/test_coverage.py:
```diff
@@ -0,0 +1,133 @@
+You should have received a copy of the GNU Affero General Public License
+along with Autometa. If not, see <http://www.gnu.org/licenses/>.
+COPYRIGHT
+Script containing wrapper functions for bedtools.
```
This needs to be updated.

Sidduppal replied:
Anything in particular I should know about that needs to be changes?

evanroyrees replied:
I was referring to "Script containing wrapper functions for bedtools."

evanroyrees commented on tests/unit_tests/test_diamond.py:
```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
```
Perhaps you should remove this test file from the PR and we can refocus on this later?

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -0,0 +1,260 @@
+    assert len(rev_string) == 14
+
+
+def test_init_kmers():
+    unique_kmers_5 = kmers.init_kmers(kmer_size=5)
+    assert type(unique_kmers_5) is dict
+    assert len(unique_kmers_5) == 512
+    unique_kmers_3 = kmers.init_kmers(kmer_size=3.0)
+    assert type(unique_kmers_3) is dict
+    assert len(unique_kmers_3) == 32
+    unique_kmers_5 = kmers.init_kmers(kmer_size="5")
+    assert type(unique_kmers_5) is dict
+    assert len(unique_kmers_5) == 512
+    unique_kmers_3 = kmers.init_kmers(kmer_size="3.0")
+    assert type(unique_kmers_3) is dict
+    assert len(unique_kmers_3) == 32
+    with pytest.raises(TypeError):
+        kmers.init_kmers(kmer_size=5.5)
+    with pytest.raises(TypeError):
+        kmers.init_kmers(kmer_size="5.5")
```
I do not think we should do any type conversion (in kmers.py) so we will not need multiple test cases here.

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -0,0 +1,260 @@
+
+
+@patch("os.path.getsize", return_value=2 * 1024 * 1024)
+def test_load(patched_file_size):
```
The checks for the `load(...)` function:

1. Mock the `pd.read_csv(...)` call with a return value of a `pd.DataFrame`.
    - ensure df is returned
2. We need to ensure a `TableFormatError` is raised.
    - ensure error is raised when format is incorrect.

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -0,0 +1,260 @@
+@patch("autometa.common.kmers.seq_counter")
+@patch("autometa.common.kmers.mp_counter")
+def test_count(pacthed_mp_counter, pacthed_seq_counter):
+    with pytest.raises(TypeError):
+        kmers.count(assembly=assembly, size=5.5)
+    with pytest.raises(TypeError):
+        kmers.count(assembly=assembly, size="5.5")
```
```suggestion
    for size in [5.5, "5.5"]:
        with pytest.raises(TypeError):
            kmers.count(assembly=assembly, size=size)
```

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -0,0 +1,260 @@
+    with pytest.raises(TypeError):
+        kmers.count(assembly=assembly, size="5.5")
+
+    out_df = kmers.count(assembly=assembly, size=5.0)
```
`size=5.0` should result in an error.

We might as well decrease the computational complexity of the counting as well

```suggestion
    out_df = kmers.count(assembly=assembly, size=3)
```

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -0,0 +1,260 @@
+        kmers.count(assembly=assembly, size="5.5")
+
+    out_df = kmers.count(assembly=assembly, size=5.0)
+    assert type(out_df) is pd.DataFrame
```
```suggestion
    assert isinstance(out_df, pd.DataFrame)
```

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -0,0 +1,260 @@
+    assert out_df.shape == (0, 512)
+
+    out_df = kmers.count(assembly=assembly, size=5, multiprocess=False)
+    assert type(out_df) is pd.DataFrame
```
```suggestion
    assert isinstance(out_df, pd.DataFrame)
```

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -0,0 +1,260 @@
+
+@patch("autometa.common.kmers.seq_counter")
+@patch("autometa.common.kmers.mp_counter")
+def test_count(pacthed_mp_counter, pacthed_seq_counter):
```
This seems like a case where we would want to [parametrize](https://docs.pytest.org/en/stable/parametrize.html "parametrize tests in pytest") this test. e.g. size and multiprocess as well as expected number of columns corresponding to size.

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -0,0 +1,260 @@
+    out_df = kmers.count(assembly=assembly, size=5, multiprocess=False)
+    assert type(out_df) is pd.DataFrame
+    assert out_df.index.name == "contig"
+    assert out_df.shape == (0, 512)
```
See comment about [parametrization](https://docs.pytest.org/en/stable/parametrize.html "parametrization in pytest") above. `512` here could be the expected number of kmer columns corresponding to the passed in size parameter.

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -0,0 +1,260 @@
+    assert out_df.index.name == "contig"
+    assert out_df.shape == (0, 512)
+
+    assert pacthed_mp_counter.call_count == 1
+    assert pacthed_seq_counter.call_count == 1
```
We should have respective tests for `mp_counter` and `seq_counter` in this file. Equally important that we ensure these work since they provide main functionality of this function.

evanroyrees commented on tests/unit_tests/test_metagenome.py:
```diff
@@ -1,89 +1,92 @@
-    # assert taxa_df.shape == (3587, )
+    nucls_fp, prots_fp = init_Metagenome.call_orfs()
+    assert patched_progial_run.call_count == 1
+    assert type(nucls_fp) is str
+    assert type(prots_fp) is str
```
```suggestion
    assert isinstance(nucls_fp, str)
    assert isinstance(prots_fp, str)
```

evanroyrees commented on tests/unit_tests/test_prodigal.py:
```diff
@@ -0,0 +1,120 @@
+        assembly, nucls_out=nucls_out, prots_out=prots_out, force=True, parallel=True,
+    )
+
+    assert patched_annotate_parallel.called is True
+    assert patched_annotate_sequential.called is True
+    assert patched_file_size.called is True
+    # assert patched_file_exists.called is True
+    assert patched_open.called is True
```
The comparisons using `is` and `==` are not quite the same. See [here](https://www.tutorialspoint.com/difference-between-and-is-operator-in-python#:~:text=is%20and%20equals(%3D%3D),two%20variables%20are%20the%20same. "tutorials point page on difference between the is and == operator") for more details.

```suggestion
    assert patched_annotate_parallel.called == True
    assert patched_annotate_sequential.called == True
    assert patched_file_size.called == True
    # assert patched_file_exists.called == True
    assert patched_open.called == True
```

Sidduppal committed:
Delete test_diamond.py

Sidduppal commented:
Thanks for the input, I'll make the requested changes soon. One of the questions that comes to my mind just of the bat is that, is there any particular reason why we don't want to do convert the input into an `int` at places when it is required, such as when the user inputs 4.0 or "4" or "4.0". 
It seems like a repeated problem we are having with a number of scripts. I understand that it is a fairly simple thing to ask from the user, but it seems this issue keeps coming back (like in `ncbi.py` that led us to create a [new function](https://github.com/KwanLab/Autometa/blob/79869f9ac94e59e659eaadc52fb1e3eec16b6706/autometa/taxonomy/ncbi.py#L396))
As an alternative I would also look into type hints.

Sidduppal committed:
comitting before pulling

Sidduppal committed:
pulling from WiscEvan CI/CD

Sidduppal committed:
Modified makefile

Sidduppal committed:
pulling WiscEvan CI/CD

Sidduppal committed:
pulling WiscEvan CI/CD

Sidduppal committed:
Pulling from WiscEvan CI/CD

Sidduppal committed:
Merge branch 'CI_CD' of https://github.com/Sidduppal/Autometa into CI_CD

Sidduppal committed:
:green_heart: Add test for recirsive_dbscan.py
:green_heart: add .travis.yaml file

Sidduppal commented:
I have also added the `.travis.yaml` file. Since, it is not my repository I cannot see the build, however, I think you can by going to [Travis](https://travis-ci.org/)

evanroyrees commented on .travis.yaml:
```diff
@@ -0,0 +1,8 @@
+language: python
+python:
+  - "2.7"
```
FYI, We are not supporting 2.7

Sidduppal committed:
removing 2.7 from .travis.yaml

Sidduppal committed:
Merge branch 'CI/CD' of https://github.com/WiscEvan/Autometa into CI_CD

Sidduppal committed:
Updated tests for recursive dbscan
Updated a few lines in vote.py
Included sam, bam and bed files in make_test_data.py
Updated .pre-commit-config.yaml file to make sure the commit hooks work on all versions of pythion and not just 3.7

Sidduppal committed:
Changes to make_test_data.py

Sidduppal committed:
:white_check_mark: Added tests for
:white_check_mark: Updated tests for
:art: Updated parsing of alignment files to

Sidduppal committed:
Merging from upstream

Sidduppal committed:
:green_heart: Added tests for coverage.py
:art: Removed bug from bowtie.py
:art: updated make_test_data.py to use fwd and rev reads

Sidduppal committed:
Pulling from upstream

Sidduppal committed:
Pulling from upstream

Sidduppal committed:
:green_heart: Added unit tests for coverage.py

Sidduppal committed:
:green_heart: test for argpase block of coverage.py, metagenome.py and kmers.py
:bug: Resolved a bug in metagenome.py
:art: Modified the code for writing tests for arparse

evanroyrees commented on autometa/common/metagenome.py:
```diff
@@ -144,7 +144,8 @@ def seqrecords(self):
 
         Returns
-        -------
+
+-------
```
This looks like you have accidentally deleted some of the docstring?

evanroyrees commented on .travis.yaml:
```diff
@@ -0,0 +1,7 @@
+  - "3.7"
+  - "3.8"
+# command to run tests
+script:
+  - pytest
```
The command we have been using is `python -m pytest`

I'm not sure if this should be replaced by this or not

evanroyrees commented on autometa/common/external/bowtie.py:
```diff
@@ -128,7 +128,8 @@ def align(db, sam, fwd_reads=None, rev_reads=None, se_reads=None, cpus=0, **kwar
         if reads:
             reads_provided = True
-            params.append(f'{flag} {",".join(reads)}')
+            # params.append(f'{flag} {",".join(reads)}')
+            params.append(f"{flag} {reads}")
```
This should be one or the other. No need to leave the commented out line

Sidduppal replied:
I wasn't sure why you had put `{",".join(reads)}` I faced an error with this while running the test and removing `","` resolved it. However, I wasn't sire if this was the right approach as you must have put that for a reason.

evanroyrees replied:
`reads` is a `List[str]`, so when emitting the message to the user I wanted to concatenate the list by comma delimiters.

Perhaps this is an issue with the quotes? 

You could try:

```python
params.append(f"{flag} {', '.join(reads)}")
```

or:

```python
reads_list = ", ".join(reads)}
params.append(f"{flag} {reads_list}")

```

Sidduppal replied:
I am not sure why you want to concatenate the using a comma. `reads` tend to become a `str` when it is being looped through by the `for` loop. This is what I was experiencing when running it in an IDE:
```python
In [12]: fwd_reads = "path_to_fwd_reads"

In [13]: rev_reads = "path_to_rev_reads"

In [14]: se_reads="path_to_se_reads"

In [15]: for flag, reads in zip(["-1", "-2", "-U"], [fwd_reads, rev_reads, se_reads]):
    ...:     if reads:
    ...:         print(",".join(reads)+"\n")
    ...:         print(', '.join(reads)+"\n")
    ...: 
    ...: 
p,a,t,h,_,t,o,_,f,w,d,_,r,e,a,d,s

p, a, t, h, _, t, o, _, f, w, d, _, r, e, a, d, s

p,a,t,h,_,t,o,_,r,e,v,_,r,e,a,d,s

p, a, t, h, _, t, o, _, r, e, v, _, r, e, a, d, s

p,a,t,h,_,t,o,_,s,e,_,r,e,a,d,s

p, a, t, h, _, t, o, _, s, e, _, r, e, a, d, s
```

evanroyrees commented on autometa/taxonomy/vote.py:
```diff
@@ -186,6 +186,7 @@ def majority_vote_lca(outfpath=outfpath):
     for fp, argname in zip(
         [lca_fpath, hits, blast, prot_orfs], ["lca", "orfs", "orfs", "orfs"],
     ):
+        print(fp)
```
We should not have any print statements

evanroyrees commented on make_test_data.py:
```diff
@@ -239,14 +245,64 @@ def get_taxonomy(self, num_orfs: int = 2):
+        bed = self.get_bed_alignments()
+        sam = self.get_sam_alignments()
+        logging.info("Getting fwd and rev reads ...")
+        reads = self.get_reads()
+        fwd_reads = reads[0]
+        rev_reads = reads[1]
```
Let's do tuple unpacking here to keep retrieval of any piece of test data to one line

```suggestion
        fwd_reads, rev_reads = self.get_reads()
```

evanroyrees commented on make_test_data.py:
```diff
@@ -239,14 +245,64 @@ def get_taxonomy(self, num_orfs: int = 2):
+        fwd_reads = reads[0]
+        rev_reads = reads[1]
+        self.data["coverage"] = {
+            "spades_records": self.metagenome,
```
We do not want to duplicate data in our test set. This is adding redundancy and unnecessarily increasing our file size.
```suggestion
```

evanroyrees commented on make_test_data.py:
```diff
@@ -58,8 +58,10 @@
 import gzip
 import json
-
+import subprocess
```
```suggestion

```

evanroyrees commented on make_test_data.py:
```diff
@@ -58,8 +58,10 @@
 import json
-
+import subprocess
+import os
```
I don't see the use of `os` here either.

```suggestion

```

evanroyrees commented on tests/unit_tests/test_coverage.py:
```diff
@@ -20,10 +23,62 @@ def fixture_metagenome(variables, tmp_path):
+    return fpath.as_posix()
+
+
+@pytest.fixture(name="bam_alignment")
+def fixture_alignment_bam(variables, sam_alignment, tmp_path):
+    bam_fpath = tmp_path / "records.bam"
+    samtools.sort(sam=sam_alignment, bam=bam_fpath)
+    return bam_fpath.as_posix()
```
Setting up a fixture with an external dependency seems a bit precarious. We could leave this for now (assuming this works) but I can imagine a case where somewhere the external `samtools.sort` gets broken and now this fixture is broken for the rest of the coverage test suite. It seems like we should separately test `samtools.sort` in `test_samtools.py` and monkeypatch the calls to each of the external tools.

Sidduppal replied:
I agree that in case `samtool.sort` gets broken it will disrupt the entire fixture. However, this is the only way I could find to write a `bam` file in the most memory-efficient manner. Any other suggestions?

evanroyrees commented on tests/unit_tests/test_coverage.py:
```diff
@@ -34,21 +89,116 @@ def test_coverage_get_from_spades(metagenome, tmp_path):
+    df = coverage.get(
+        fasta=metagenome,
+        from_spades=False,
+        out=out.as_posix(),
```
Looking into Pathlib methods/operators [:link:](https://docs.python.org/3/library/pathlib.html#operators "Pathlib operators") it seems we should be using `str(out)` here instead of `out.as_posix()`

```suggestion
        out=str(out),
```

This should be applied across _all_ instances of `PathWeHaveDefined.as_posix()` -> `str(PathWeHaveDefined)`.

Sidduppal replied:
It seems we don't asctually need `str` as just inputing the `posix path` is working as well.

Sidduppal replied:
Also, I found a [good resource](https://www.linuxjournal.com/content/easier-python-paths-pathlib) on why using pathlib objects is actually a more python way of handling files.

evanroyrees commented on tests/unit_tests/test_coverage.py:
```diff
@@ -20,10 +23,62 @@ def fixture_metagenome(variables, tmp_path):
+    return bam_fpath.as_posix()
+
+
+@pytest.fixture(name="bed_alignment")
+def fixture_alignment_bed(variables, tmp_path):
+    coverage_test_data = variables["coverage"]
+    alignment_records = pd.read_json(coverage_test_data["bed"])
+    fpath = tmp_path / "records.bed"
+    alignment_records.to_csv(fpath, sep="\t", header=True, index=False)
+    return fpath.as_posix()
```
`bedtools.parse` uses the `pd.read_csv` method to read in the table. `pd.read_csv` can handle Pathlib objects. So I'm not entirely sure, but I do not think you need to use `fpath.as_posix()` but could simply return `fpath`. 
 Otherwise, see comments below about `fpath.as_posix()` -> `str(fpath)`

evanroyrees commented on tests/unit_tests/test_coverage.py:
```diff
@@ -34,21 +89,116 @@ def test_coverage_get_from_spades(metagenome, tmp_path):
+    coverage.get(fasta=metagenome, out=df_exists_fpath, bed=bed_alignment)
+
+
+use_spades = [True, False]
```
This either needs to be removed or employed somewhere.
```suggestion

```

evanroyrees commented on tests/unit_tests/test_coverage.py:
```diff
@@ -34,21 +89,116 @@ def test_coverage_get_from_spades(metagenome, tmp_path):
+                out,
+            )
+
+    # Defining the MockParser class to represent parser
```
`# Defining the MockParser class to represent parser`

⬆️  This comment should be above `class MockParser`

`monkeypatch.setattr(argparse, "ArgumentParser", return_mock_parser, raising=True)` is not defining the `MockParser` class, but patching `argparse.ArgumentParser()` with a function `return_mock_parser`.

evanroyrees commented on tests/unit_tests/test_coverage.py:
```diff
@@ -34,21 +89,116 @@ def test_coverage_get_from_spades(metagenome, tmp_path):
+use_spades = [True, False]
+
+
+@pytest.fixture(name="mock_parser", params=use_spades)
+def fixture_mock_parser(
+    metagenome,
+    sam_alignment,
+    bam_alignment,
+    bed_alignment,
+    fwd_reads,
+    rev_reads,
+    monkeypatch,
+    tmp_path,
+    request,
+):
+    def return_mock_parser(*args, **kwargs):
```
As far as I can tell this does not need to be a fixture but can all be defined _within_ the `test_coverage_main` function, No?

evanroyrees replied:
Also, there is no need to add all of these other fixtures to this. We just want to ensure the main logic is appropriate. We do _not_ want to rerun all of the functions together since we already have separate tests for them. This may be adding unnecessary time to our tests.

evanroyrees commented on tests/unit_tests/test_metagenome.py:
```diff
@@ -162,7 +164,41 @@ def test_orfs_called(metagenome, monkeypatch):
     with monkeypatch.context():
         # Here we set the filepaths to the metagenome fixture assembly path
         # (since we know this exists and is non-empty)
+        # metagenome.__str__ == metagenome.assembly
         monkeypatch.setattr(metagenome, "prot_orfs_fpath", str(metagenome))
         monkeypatch.setattr(metagenome, "nucl_orfs_fpath", str(metagenome))
```
This is more explicit. I should have aliased the context in the first place

```suggestion
    with monkeypatch.context() as m:
        # Here we set the filepaths to the metagenome fixture assembly path
        # (since we know this exists and is non-empty)
        # metagenome.__str__ == metagenome.assembly
        m.setattr(metagenome, "prot_orfs_fpath", str(metagenome))
        m.setattr(metagenome, "nucl_orfs_fpath", str(metagenome))
```

evanroyrees commented on tests/unit_tests/test_metagenome.py:
```diff
@@ -1,4 +1,6 @@
 import pytest
+import os
```
Is `os` being used in this script?

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -158,3 +185,64 @@ def test_embed_out_exists(norm_df, tmp_path):
+        kmers.embed(kmers=kmer_fpath)
+
+
+@patch("os.path.getsize", return_value=2 * 1024 * 1024)
```
I think we should not use multiple different patching and test methods/frameworks. This adds complexity to the testing suite and forces other developers to learn more frameworks than is necessary. It is not clear what this patch is doing or why it is necessary.

```suggestion

```

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -1,7 +1,13 @@
 from Bio import SeqIO
+
 from autometa.common import kmers
+from unittest.mock import patch, MagicMock
```
Let's stick to only one testing framework to reduce confusion, complexity, etc.

```suggestion

```

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -1,7 +1,13 @@
 import pytest
+import os
```
Is `os` being used in this?

evanroyrees commented on tests/unit_tests/test_prodigal.py:
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python
```
I think it may be best to scrap this and start from scratch using pytest

evanroyrees commented on tests/unit_tests/test_recursive_dbscan.py:
```diff
@@ -12,7 +14,7 @@ def fixture_kmers(variables, tmp_path):
     fpath = tmp_path / "kmers.norm.tsv"
     df.to_csv(fpath, sep="\t", index=False, header=True)
-    return fpath
+    return fpath.as_posix()
```
This is incorrect, see comments above.

evanroyrees commented on tests/unit_tests/test_recursive_dbscan.py:
```diff
@@ -104,6 +116,20 @@ def test_binning_invalid_clustering_method(master, markers):
         )
 
 
+@pytest.mark.wip
+def test_binning_empty_markers_table(
+    df_without_markers, master,
+):
+    with pytest.raises(TableFormatError):
+        recursive_dbscan.binning(
+            master=master,
+            markers=df_without_markers,
+            domain="bacteria",
+            method="hdbscan",
+            taxonomy=False,
+        )
```
We may be over using fixture here. Since this `df_without_markers` fixture is only used once, we can just define it within this test function. For one, this enhances readability and two, we can try to prevent any unintended side effects that may be caused by any particular fixture.

evanroyrees commented on tests/unit_tests/test_vote.py:
```diff
@@ -180,6 +180,7 @@ def test_add_ranks(ncbi_dir, votes, tmp_path):
         assert canonical_rank in df.columns
 
 
+@pytest.mark.skip
```
This should be removed.
```suggestion

```

evanroyrees commented on tests/unit_tests/test_recursive_dbscan.py:
```diff
@@ -104,6 +116,20 @@ def test_binning_invalid_clustering_method(master, markers):
         )
 
 
+@pytest.mark.wip
```
Is this still a WIP? It doesn't appear to be so.

```suggestion

```

evanroyrees commented on tests/unit_tests/test_samtools.py:
```diff
@@ -0,0 +1,43 @@
+sam_fpath = "<path/to/mock/sam/file>"
+
+
+def test_sort():
```
The name of this test function should be more descriptive.

evanroyrees reviewed (changes requested):
I've placed comments throughout. It seems you have littered some formatting issues regarding docstrings and print statements. There is a key difference in how you are handling Pathlib objects that needs to be addressed. You should also remove any imports that are not used in the respective script. We should move away from using anything from `unittest` as majority of the tests are written with `pytest` and we are running `pytest` to perform our testing (i.e MagicMocks and Mocks should be discarded). prodigal is not using any test data should probably be scrapped to rework from scratch using fixtures and monkeypatching as we have done with the other test scripts. There is a subtlety in the `argparse.ArgumentParser()` monkeypatch when calling main for our entrypoints. From `test_recursive_dbscan.py` you can see the file paths passed in and checked as keyword arguments against the monkeypatched `recursive_dbscan.binning()` method (this is called in main and passed args). This way we are ensuring the argparse logic is passing the appropriate arguments to the logic we have laid out in main. This should similarly be done for our other entrypoints. Right now you are setting up the mocks as a fixture and only calling main. We need to test the logic, _not_ the functions within main. (These are tested separately in the test script)

Overall I think we are making great progress and it appears to me that you are getting the hang of things 🛠️ 👷  

We are almost there ✅

Sidduppal committed:
Pulling from upstream

Sidduppal committed:
Addressed code review comments

Sidduppal committed:
skip metagenome main test

Sidduppal committed:
:art: :green-heart: Modified .travis.yml to run autometa environment and install all the dependencies

Sidduppal committed:
:art: :green-heart: Modified .travis.yml to include codecov

Sidduppal committed:
Update .travis.yml

:green_heart: :art: miniconda update has renamed the default installation path to miniconda2
See [:link:](https://stackoverflow.com/a/34257781/12671809)

Sidduppal committed:
Update .travis.yml

Made changes as per the official documentation - [link](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/use-conda-with-travis-ci.html#the-travis-yml-file)

Sidduppal committed:
Update .travis.yml

Sidduppal committed:
Removed ` bash miniconda.sh -b -p $HOME/miniconda`

Sidduppal committed:
Downloading miniconda3 

:art: :green_heart:  Added `chmod +x miniconda.sh`

Sidduppal committed:
:green_heart: :art: Moved argparse ficture inside testing_main() for coverage and metagenome
:art: changes session scopes for metagenome and coverage

Sidduppal committed:
Merge branch 'CI_CD' of https://github.com/Sidduppal/Autometa into CI_CD

evanroyrees commented on autometa/common/external/bowtie.py:
```diff
@@ -128,7 +128,7 @@ def align(db, sam, fwd_reads=None, rev_reads=None, se_reads=None, cpus=0, **kwar
         if reads:
             reads_provided = True
-            params.append(f'{flag} {",".join(reads)}')
+            params.append(f"{flag} {reads}")
```
This is just a formatting nitpick.

```suggestion
            listed_reads = ", ".join(reads)
            params.append(f"{flag} {listed_reads}")
```

Sidduppal committed:
pulling from WiscEvan upstream

Sidduppal committed:
:art: Formatting make_test_data.py

evanroyrees commented on autometa/common/external/prodigal.py:
```diff
@@ -186,7 +186,7 @@ def run(assembly, nucls_out, prots_out, force=False, cpus=0, parallel=True):
                     pass
-        except (IOError, ValueError):
+        except (OSError, ValueError):
             raise IOError(f"InvalidFileFormat: {fp}")
```
```suggestion
            raise OSError(f"{fp} file is not properly formatted")
```

evanroyrees commented on autometa/common/kmers.py:
```diff
@@ -504,7 +506,7 @@ def embed(
         try:
             df = pd.read_csv(kmers, sep="\t", index_col="contig")
-        except ValueError:
+        except (ValueError):
```
```suggestion
        except ValueError:
```

evanroyrees commented on autometa/common/kmers.py:
```diff
@@ -514,13 +516,13 @@ def embed(
         try:
             return pd.read_csv(out, sep="\t", index_col="contig")
-        except ValueError:
+        except (ValueError):
```
```suggestion
        except ValueError:
```

evanroyrees commented on make_test_data.py:
```diff
@@ -267,14 +273,61 @@ def get_taxonomy(self, num_orfs: int = 2):
+        # Get number of unique contigs set by `num_contigs`, default is 1.
+        contigs = set(coverages.index.unique().tolist()[:num_contigs])
+        coverages = coverages.loc[contigs]
+        coverages.reset_index(
+            inplace=True
+        )  # Here we are ready to send to json object self.data
```
```suggestion
        coverages.reset_index(inplace=True)
        # Here we are ready to send to json object self.data
```

evanroyrees commented on make_test_data.py:
```diff
@@ -267,14 +273,61 @@ def get_taxonomy(self, num_orfs: int = 2):
             "names": names,
         }
 
+    def get_bed_alignments(self, num_contigs=1):
+        contig_col = 0
+        coverages = pd.read_csv(
+            self.bed_fpath, sep="\t", index_col=contig_col, header=None
+        )
+        # Get number of unique contigs set by `num_contigs`, default is 1.
+        contigs = set(coverages.index.unique().tolist()[:num_contigs])
+        coverages = coverages.loc[contigs]
```
```suggestion
    def get_bed_alignments(self, num_contigs=1):
        contig_col = 0
        coverages = pd.read_csv(
            self.bed_fpath, sep="\t", index_col=contig_col, header=None
        )
        # Get number of unique contigs set by `num_contigs`, default is 1.
        coverages = coverages.sample(n=num_contigs, random_state=self.seed)
```

evanroyrees commented on make_test_data.py:
```diff
@@ -267,14 +273,61 @@ def get_taxonomy(self, num_orfs: int = 2):
+                    contig = line.split("\t")[1]
+                    contig = contig.replace("SN:", "")
+                if line.startswith("@HD") or line.startswith("@PG") or contig in line:
+                    # @HD and @PG is required for bam construction
```
```suggestion
                    # @HD and @PG are required for bam construction
```

evanroyrees commented on tests/unit_tests/test_coverage.py:
```diff
@@ -1,61 +1,190 @@
+    assert out.exists()
+
+
+def test_get_ValueError(metagenome, tmp_path):
```
```suggestion
def test_get_not_enough_passed_arguments(metagenome, tmp_path):
```

evanroyrees commented on tests/unit_tests/test_coverage.py:
```diff
@@ -1,61 +1,190 @@
-    out = tmp_path / "covs_from_spades.tsv"
-    df = coverage.get(fasta=metagenome, from_spades=True, out=out)
+@pytest.fixture(name="bam_alignment", scope="module")
+def fixture_bam_alignment(variables, sam_alignment, coverage_testdir):
```
```suggestion
def fixture_bam_alignment(sam_alignment, coverage_testdir):
```

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -46,12 +48,34 @@ def fixture_norm_df(variables):
     assert df.index.name == "contig"
 
 
+def test_kmer_load_FileNotFoundError():
```
```suggestion
def test_kmer_load_missing_file():
```

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -46,12 +48,34 @@ def fixture_norm_df(variables):
+        kmers.load(kmers_fpath="Invalid_fpath")
+
+
+def test_kmer_load_TableFormatError(invalid_df_fpath):
```
```suggestion
def test_kmer_load_table_missing_contig_column(invalid_df_fpath):
```

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -170,6 +193,24 @@ def test_embed_out_exists(norm_df, force, tmp_path):
     )
 
 
+def test_embed_TableFormatError(invalid_df_fpath):
```
```suggestion
def test_embed_table_missing_contig_column(invalid_df_fpath):
```

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -170,6 +193,24 @@ def test_embed_out_exists(norm_df, force, tmp_path):
+        kmers.embed(kmers=kmer_fpath)
+
+
+def test_embed_FileNotFoundError(tmp_path):
```
```suggestion
def test_embed_empty_dataframe(tmp_path):
```

evanroyrees commented on tests/unit_tests/test_kmers.py:
```diff
@@ -170,6 +193,24 @@ def test_embed_out_exists(norm_df, force, tmp_path):
+        kmers.embed(kmers=invalid_df_fpath)
+
+
+def test_embed_TypeError(tmp_path):
```
```suggestion
def test_embed_input_not_string_or_dataframe(tmp_path):
```

evanroyrees commented on tests/unit_tests/test_metagenome.py:
```diff
@@ -160,17 +165,67 @@ def test_call_orfs_invalid_params(metagenome, force, parallel, cpus):
+        m.setattr(metagenome, "nucl_orfs_fpath", str(metagenome))
+
+
+@pytest.fixture(name="mock_parser")
+def fixture_mock_parser(
+    assembly, monkeypatch, tmp_path,
+):
+    def return_mock_parser(*args, **kwargs):
+        return MockParser()
+
+    class MockParseArgs:
+        def __init__(self, assembly, out):
+            self.assembly = assembly
+            self.force = True
+            self.cutoff = 3000
+            self.out = out
+            self.stats = True
+
+    # Defining the MockParser class to represent parser
+    class MockParser:
+        def add_argument(self, *args, **kwargs):
+            pass
+
+        def parse_args(self):
+            out = tmp_path / "binning.tsv"
+            return MockParseArgs(assembly, out)
+
+    monkeypatch.setattr(argparse, "ArgumentParser", return_mock_parser, raising=True)
```
```suggestion
```

evanroyrees commented on tests/unit_tests/test_recursive_dbscan.py:
```diff
@@ -126,7 +128,22 @@ def test_binning_invalid_clustering_method(master, markers):
 
 
-@pytest.mark.wip
+def test_binning_empty_markers_table(master,):
```
```suggestion
def test_binning_empty_markers_table(master):
```

evanroyrees commented on tests/unit_tests/test_samtools.py:
```diff
@@ -0,0 +1,43 @@
+from autometa.common.external import samtools
+
+
+sam_fpath = "<path/to/mock/sam/file>"
+
+
+def test_samtools_sort():
+
+    with pytest.raises(FileNotFoundError):
+        samtools.sort(sam="sam", bam="bam")
+    with pytest.raises(TypeError):
+        samtools.sort(sam="sam_fpath", bam="bam", cpus=2.9)
+    with pytest.raises(TypeError):
+        samtools.sort(
+            sam="sam_fpath", bam="bam", cpus=-2,
+        )
```
```suggestion
def test_sort_missing_file():
    with pytest.raises(FileNotFoundError):
        samtools.sort(sam="sam", bam="bam")

@pytest.mark.parametrize("cpus", [2.9, -2])
def test_sort_invalid_cpu_input(cpus):
    with pytest.raises(TypeError):
        samtools.sort(sam="sam_fpath", bam="bam", cpus=cpus)
```

evanroyrees committed:
Apply suggestions from code review
